vivarium/core/registry.py

```diff
@@ -335,130 +363,73 @@ def divide_null(state):
+    method is called instead and will be much slower.
 
     Args:
         name: Name of the serializer. Defaults to the class name.
-        exclusive_types: Types that are exclusively handled by this
-            serializer. These can be used to quickly lookup which
-            serializer handles a particular type. If an object does not
-            match any of these exclusive types, the `can_serialize`
-            method will be used as a fallback.
-
-            Note that exclusive type matching is exact--inheritance is
-            not accounted for. This means that if your serializer has an
-            exclusive type of A and B inherits from A, an instance of B
-            will not match to your serializer's exclusive types.
     """
     REGEX_FOR_NAME = re.compile('[A-Za-z0-9-_]+')
     REGEX_FOR_SERIALIZED_ANY_TYPE = re.compile(
         f'!{REGEX_FOR_NAME.pattern}\\[(.*)\\]')
 
-    def __init__(self, name='', exclusive_types=tuple()):
+    def __init__(self, name=''):
         self.name = name or self.__class__.__name__
         self.regex_for_serialized = re.compile(f'!{self.name}\\[(.*)\\]')
-        self.alternate_keys = (
-            str(exclusive_type)
-            for exclusive_type in exclusive_types
-        )
-
-    def serialize_to_string(self, data):
-        """Serialize some data.
-
-        Subclasses should override this function,
-
-        Args:
-            data: Data to serialize.
-
-        Returns:
-            The serialized data.
+        self.codecs = self.get_codecs()
+        for codec in self.codecs:
+            if hasattr(codec, 'bson_type') and hasattr(codec, 'transform_bson'):
+                self.deserialize = lambda x: x
+                self.can_deserialize = lambda _: False
+    
+    def get_codecs(self):
+        """Get list of codecs in serializer. Codecs are class attributes of type
+        :py:class:`bson.codec_options.TypeCodec` that are used by PyMongo to
+        serialize and (optionally) deserialize data.
         """
-        raise NotImplementedError(
-            f'{self} has not implemented serialize_to_string().')
-
+        return []
+    
     def serialize(self, data):
-        """Serialize some data.
-
-        Subclasses that are serializing to strings can just implement
-        :py:meth:`vivarium.core.registry.Serializer.serialize_to_string`,
-        and this method will use that implementation to generate a
-        string that serializes the data.
-
-        Subclasses with more complex serialization needs must override
-        this function.
-
-        Args:
-            data: Data to serialize.
-
-        Returns:
-            The serialized data.
+        """Serialize data using correct codec.
+        
+        This is typically only called in the case that individual stores
+        are assigned custom serializers. For maximum performance, serialization
+        should be left to PyMongo instead of calling this function.
```

Comment:
ditto: explain if/when this function should be overridden by subclasses

Reply:
As is, this method works seamlessly with the unit conversion logic in the if block beginning on line 1022 of `store.py`. 

This should only ever be overriden in the case described, when individual stores are assigned to the serializer subclass via the `_serializer` key. For performance reasons, the `_serializer` key should only be defined in the case that the data in that store is of a type that either can't be serialized by a custom codec (i.e. built-ins) or already has a defined codec but needs to be serialized differently from other instances of that type. In that case, users can override this `serialize` method to define whatever serialization routine they desire. 

Note that the values returned by this method later undergo another round of serialization under PyMongo's codec system, potentially causing unexpected behavior if `serialize` returns data of a type that is modified by another codec before insertion into the database. Best practice would be for `serialize` to always return data of a built-in type with predictable serialization behavior that cannot be modified by the end user. 

I'll update the docstring with these details.